packages/core/src/conflict-resolution.ts

```diff
@@ -172,59 +172,51 @@ export class HistoryLog {
  * @param dispatcher - Get commit from IPFS
  * @param cid - Commit CID
  * @param stateLog - Log from the current stream state
- * @param log - Found log so far
+ * @param unappliedCommits - Unapplied commits found so far
  * @param timestamp - Previously found timestamp
  * @private
  */
 export async function fetchLog(
   dispatcher: Dispatcher,
   cid: CID,
   stateLog: HistoryLog,
-  log: LogEntry[] = [],
+  unappliedCommits: CommitData[] = [],
   timestamp?: number
-): Promise<LogEntry[]> {
+): Promise<CommitData[]> {
   if (stateLog.includes(cid)) {
     // already processed
     return []
   }
   const commit = await dispatcher.retrieveCommit(cid)
   if (!commit) throw new Error(`No commit found for CID ${cid.toString()}`)
 
-  let payload = commit
-  const isSignedCommit = StreamUtils.isSignedCommit(commit)
-  if (isSignedCommit) {
-    payload = await dispatcher.retrieveCommit(commit.link)
-    if (!payload) throw new Error(`No commit found for CID ${commit.link.toString()}`)
-  }
-  const prevCid: CID = payload.prev
-  if (!prevCid) {
-    // this is a fake log
-    return []
-  }
-  let nextEntry: LogEntry
-  if (StreamUtils.isAnchorCommit(commit)) {
+  let nextCommitData: CommitData
+  if (StreamUtils.isSignedCommit(commit)) {
+    const linkedCommit = await dispatcher.retrieveCommit(commit.link)
+    if (!linkedCommit) throw new Error(`No commit found for CID ${commit.link.toString()}`)
+    nextCommitData = { cid: cid, type: CommitType.SIGNED, commit: linkedCommit, envelope: commit, timestamp: timestamp }
+  } else if (StreamUtils.isAnchorCommit(commit)) {
     const proof = await dispatcher.retrieveFromIPFS(commit.proof)
     timestamp = proof.blockTimestamp
-    nextEntry = {
-      type: CommitType.ANCHOR,
-      cid: cid,
-      timestamp: timestamp,
-    }
+    nextCommitData = { cid: cid, type: CommitType.ANCHOR, commit: commit, timestamp: timestamp }
```

Comment:
One optimization (maybe for later?) we could do here is to also retrieve the merkle wittness and put it into the `nextCommitData` object.

Reply:
Cool, yes 👍🏼

Reply:
Mind creating a ticket to track this?

Reply:
Done #1544